contractcourt/commit_sweep_resolver.go

```diff
@@ -262,22 +267,30 @@ func (c *commitSweepResolver) Resolve() (ContractResolver, error) {
 
 			return nil, sweepResult.Err
 		}
+
+		sweepTxID = sweepResult.Tx.TxHash()
+
 	case <-c.quit:
 		return nil, errResolverShuttingDown
 	}
 
 	// Funds have been swept and balance is no longer in limbo.
 	c.reportLock.Lock()
-	if recovered {
+	if outcome == channeldb.ResolverOutputCommitSweep {
 		// We only record the balance as recovered if it actually came
 		// back to us.
 		c.currentReport.RecoveredBalance = c.currentReport.LimboBalance
 	}
 	c.currentReport.LimboBalance = 0
 	c.reportLock.Unlock()
-
+	report := c.currentReport.resolverReport(&sweepTxID, outcome)
 	c.resolved = true
-	return nil, c.Checkpoint(c)
+
+	// Checkpoint the resolver with a closure that will write the outcome
+	// of the resolver and its sweep transaction to disk.
+	return nil, c.Checkpoint(c, func(tx kvdb.RwTx) error {
+		return c.PutResolverReport(tx, report)
```

Comment:
All of the resolvers now end with this `PutResolverReport` call. I'd try to move that to the caller of `Resolve` so that it isn't duplicated and always guaranteed to be executed. If you unify resolver report and contract report, the caller can request the report through the existing `report()` method.

Reply:
We want the `PutReport` call to be called in the same tx as our final `Checkpoint` (otherwise we could checkpoint, restart and then won't resume the complete resolver but also won't have the report). I'm unsure whether it makes sense to move the final `Checkpoint` out of each resolver? The came could have been said of `Checkpoint` itself before we started writing reports, but it leaves `Resolve` incomplete to move it out imo.

Reply:
I would prefer moving the `CheckPoint` out from the resolver rather than complicating the resolver with the closure.